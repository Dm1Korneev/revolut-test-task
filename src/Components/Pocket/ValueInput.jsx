import React from 'react';
import PropTypes from 'prop-types';

import { floorPlus } from 'Helpers';

const defaultProps = {
  valuePrefix: '',
  value: 0,
};

const propTypes = {
  changeValue: PropTypes.func.isRequired,
  value: PropTypes.number,
  valuePrefix: PropTypes.string,
};

function ValueInput(props) {
  const {
    changeValue, value, valuePrefix,
  } = props;

  const valueChangeHandler = (event) => {
    const { valueAsNumber, value: valueFromInput } = event.target;
    if (valueFromInput === '') {
      changeValue(null);
    }
    if (Number.isNaN(valueAsNumber)) {
      return;
    }
    let newValue = valueAsNumber < 0 ? -valueAsNumber : valueAsNumber;
    newValue = floorPlus(newValue);
    changeValue(newValue);
  };

  const valueForInput = value === null ? '' : value;
  const valuePrefixString = value === 0 ? '' : valuePrefix;

  return (
    <div>
      {valuePrefixString}
      <input type="number" value={valueForInput} step="0.01" onChange={valueChangeHandler} />
    </div>
  );
}

ValueInput.propTypes = propTypes;
ValueInput.defaultProps = defaultProps;

export default ValueInput;
