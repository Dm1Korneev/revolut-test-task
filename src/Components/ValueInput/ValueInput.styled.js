
import styled from 'styled-components';

export const Input = styled.input`
    -webkit-appearance: none;
    margin: 0;
    background-color: transparent;
    width: 10rem;
    padding: 1rem;

    border-radius: 4px;
    border: 1px solid rgba(0, 0, 0, 0.23);
`;

export const PrefixContainer = styled.span`
    padding: 1rem;
`;
