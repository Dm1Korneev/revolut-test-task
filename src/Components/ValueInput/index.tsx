import React from 'react';
import PropTypes from 'prop-types';

import { floorPlus } from 'Helpers';

import { Input, PrefixContainer } from './ValueInput.styled';

const defaultProps = {
  valuePrefix: '',
  value: 0,
};

const propTypes = {
  changeValue: PropTypes.func.isRequired,
  value: PropTypes.number,
  valuePrefix: PropTypes.string,
  label: PropTypes.string.isRequired,
};

function ValueInput(props) {
  const {
    changeValue, value, valuePrefix, label,
  } = props;

  const valueChangeHandler = (event) => {
    const { valueAsNumber, value: valueFromInput } = event.target;
    if (valueFromInput === '') {
      changeValue(null);
    }
    if (Number.isNaN(valueAsNumber)) {
      return;
    }
    let newValue = valueAsNumber < 0 ? -valueAsNumber : valueAsNumber;
    newValue = floorPlus(newValue);
    changeValue(newValue);
  };

  const valueForInput = value === null ? '' : value;
  const valuePrefixString = value === 0 ? '' : valuePrefix;

  return (
    <div>
      <PrefixContainer>{valuePrefixString}</PrefixContainer>
      <Input type="number" value={valueForInput} step="0.01" onChange={valueChangeHandler} aria-label={label} />
    </div>
  );
}

ValueInput.propTypes = propTypes;
ValueInput.defaultProps = defaultProps;

export default ValueInput;
