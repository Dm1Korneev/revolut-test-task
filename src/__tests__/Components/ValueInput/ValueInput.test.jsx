import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import ValueInput from 'Components/ValueInput/';

describe('ValueInput component', () => {
  const changeValue = jest.fn();
  const props = {
    valuePrefix: '+',
    value: 100.54,
    changeValue,
    label: 'Test label',
  };

  test('renders the component', () => {
    const { asFragment } = render(<ValueInput {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('change value', () => {
    const { getByRole } = render(<ValueInput {...props} />);
    fireEvent.change(getByRole('textbox'), { target: { value: '200.01', valueAsNumber: 200.01 } });
    expect(changeValue).toHaveBeenCalledWith(200.01);
  });
});
