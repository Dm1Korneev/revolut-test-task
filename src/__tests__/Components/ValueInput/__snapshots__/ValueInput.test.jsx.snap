// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValueInput component renders the component 1`] = `
<DocumentFragment>
  <div>
    <span
      class="sc-bwzfXH hnmbcI"
    >
      +
    </span>
    <input
      aria-label="Test label"
      class="sc-bdVaJa goBKDi"
      step="0.01"
      type="number"
      value="100.54"
    />
  </div>
</DocumentFragment>
`;
